File: src/leash/policy/load.py
"""
Policy parsing and validation.

Validation is strict and rejects unknown keys. In a security policy a typo
is not a harmless no-op: ``startWith`` silently ignored would turn a scoped
filesystem rule into an unscoped one. Fail loudly at load time instead.
"""

from __future__ import annotations

import math
import re
from typing import Any, Final

import yaml

from .._js import js_json
from ..types import ArgConstraint, Effect, Policy, PolicyRule, RateLimit, RunPlan
from ..types import BudgetLimits


class PolicyError(Exception):
    """A policy that cannot be trusted. Carries the location of the problem."""

    def __init__(self, message: str, path: str) -> None:
        super().__init__(f"{path}: {message}")
        self.path = path


EFFECTS: Final[frozenset[str]] = frozenset({"allow", "deny", "ask"})
_POLICY_KEYS: Final = ("version", "name", "default", "budget", "plan", "rules")
_PLAN_KEYS: Final = ("purpose", "approvedBy", "warnAt")
_RULE_KEYS: Final = ("id", "description", "tools", "effect", "when", "limit")
_BUDGET_KEYS: Final = ("calls", "tokens", "usd", "seconds")
_LIMIT_KEYS: Final = ("max", "perSeconds")
_CONSTRAINT_KEYS: Final = (
    "optional", "oneOf", "noneOf", "matches", "startsWith", "excludes",
    "min", "max", "maxLength", "minLength", "urlHosts", "type",
)
_ARG_TYPES: Final = ("string", "number", "boolean", "object", "array")


def load_policy_file(file: str) -> Policy:
    """Read and validate a policy from a ``.yaml``, ``.yml`` or ``.json`` file."""
    try:
        with open(file, "r", encoding="utf-8") as handle:
            raw = handle.read()
    except OSError as err:
        raise PolicyError(f"cannot read policy file ({err.strerror})", file) from err
    return parse_policy(raw, file)


def parse_policy(raw: str, source: str = "policy") -> Policy:
    """Validate an already-read policy document. ``source`` only decorates errors."""
    try:
        doc = yaml.safe_load(raw)
    except yaml.YAMLError as err:
        raise PolicyError(f"invalid YAML/JSON ({err})", source) from err
    return validate_policy(doc, source)


def validate_policy(doc: Any, source: str = "policy") -> Policy:
    root = _require_object(doc, source)
    _reject_unknown_keys(root, _POLICY_KEYS, source)

    if root.get("version") != "1":
        raise PolicyError(
            f"unsupported policy version {js_json(root.get('version'))}; expected \"1\"",
            source,
        )

    name = root.get("name")
    if name is not None and not isinstance(name, str):
        raise PolicyError('"name" must be a string', source)

    default = root.get("default")
    if default is not None and default not in EFFECTS:
        raise PolicyError('"default" must be one of allow, deny, ask', source)

    policy = Policy(version="1", rules=[], name=name, default=default)

    budget_raw = root.get("budget")
    if budget_raw is not None:
        budget = _require_object(budget_raw, f"{source}.budget")
        _reject_unknown_keys(budget, _BUDGET_KEYS, f"{source}.budget")
        for key in _BUDGET_KEYS:
            value = budget.get(key)
            if value is None:
                continue
            if not _is_number(value) or not math.isfinite(value) or value <= 0:
                raise PolicyError(f'"{key}" must be a positive number', f"{source}.budget")
        policy.budget = BudgetLimits(**{k: budget.get(k) for k in _BUDGET_KEYS})

    plan_raw = root.get("plan")
    if plan_raw is not None:
        plan = _require_object(plan_raw, f"{source}.plan")
        _reject_unknown_keys(plan, _PLAN_KEYS, f"{source}.plan")

        purpose = plan.get("purpose")
        if not isinstance(purpose, str) or purpose.strip() == "":
            raise PolicyError('"purpose" must be a non-empty string', f"{source}.plan")
        approved_by = plan.get("approvedBy")
        if approved_by is not None and not isinstance(approved_by, str):
            raise PolicyError('"approvedBy" must be a string', f"{source}.plan")

        warn_at = plan.get("warnAt")
        if warn_at is not None:
            if not isinstance(warn_at, list) or len(warn_at) == 0:
                raise PolicyError('"warnAt" must be a non-empty array', f"{source}.plan")
            for threshold in warn_at:
                if not _is_number(threshold) or not threshold > 0 or threshold > 1:
                    raise PolicyError(
                        '"warnAt" entries must be fractions greater than 0 and at '
                        "most 1, e.g. 0.8",
                        f"{source}.plan",
                    )
        # A plan with nothing to measure against would report nothing, which reads
        # as "all clear" rather than "not configured". Fail loudly instead.
        if budget_raw is None:
            raise PolicyError(
                'a "plan" requires a "budget" to measure against; add budget limits '
                "or remove the plan",
                source,
            )

        policy.plan = RunPlan(
            purpose=purpose,
            approvedBy=approved_by,
            warnAt=sorted(warn_at) if warn_at is not None else None,
        )

    rules_raw = root.get("rules")
    if not isinstance(rules_raw, list):
        raise PolicyError('"rules" must be an array', source)

    seen: set[str] = set()
    for i, raw_rule in enumerate(rules_raw):
        parsed = _validate_rule(raw_rule, f"{source}.rules[{i}]")
        if parsed.id in seen:
            raise PolicyError(f'duplicate rule id "{parsed.id}"', f"{source}.rules[{i}]")
        seen.add(parsed.id)
        policy.rules.append(parsed)

    return policy


def _validate_rule(raw: Any, path: str) -> PolicyRule:
    rule = _require_object(raw, path)
    _reject_unknown_keys(rule, _RULE_KEYS, path)

    identifier = rule.get("id")
    if not isinstance(identifier, str) or identifier == "":
        raise PolicyError('"id" must be a non-empty string', path)
    description = rule.get("description")
    if description is not None and not isinstance(description, str):
        raise PolicyError('"description" must be a string', path)
    tools = rule.get("tools")
    if not isinstance(tools, list) or len(tools) == 0:
        raise PolicyError('"tools" must be a non-empty array of glob patterns', path)
    for tool in tools:
        if not isinstance(tool, str) or tool == "":
            raise PolicyError('every entry in "tools" must be a non-empty string', path)
    effect = rule.get("effect")
    if effect not in EFFECTS:
        raise PolicyError('"effect" must be one of allow, deny, ask', path)

    out = PolicyRule(
        id=identifier,
        tools=list(tools),
        effect=cast_effect(effect),
        description=description,
    )

    when_raw = rule.get("when")
    if when_raw is not None:
        when = _require_object(when_raw, f"{path}.when")
        out.when = {
            arg_path: _validate_constraint(constraint, f'{path}.when["{arg_path}"]')
            for arg_path, constraint in when.items()
        }

    limit_raw = rule.get("limit")
    if limit_raw is not None:
        limit = _require_object(limit_raw, f"{path}.limit")
        _reject_unknown_keys(limit, _LIMIT_KEYS, f"{path}.limit")
        maximum = limit.get("max")
        if not _is_number(maximum) or not _is_integer(maximum) or maximum < 0:
            raise PolicyError('"max" must be a non-negative integer', f"{path}.limit")
        per_seconds = limit.get("perSeconds")
        if per_seconds is not None and (not _is_number(per_seconds) or per_seconds <= 0):
            raise PolicyError('"perSeconds" must be a positive number', f"{path}.limit")
        out.limit = RateLimit(max=int(maximum), perSeconds=per_seconds)

    return out


def _validate_constraint(raw: Any, path: str) -> ArgConstraint:
    c = _require_object(raw, path)
    _reject_unknown_keys(c, _CONSTRAINT_KEYS, path)

    matches = c.get("matches")
    if matches is not None:
        if not isinstance(matches, str):
            raise PolicyError('"matches" must be a string', path)
        try:
            re.compile(matches)
        except re.error as err:
            raise PolicyError(f'"matches" is not a valid regexp ({err})', path) from err

    for key in ("startsWith", "excludes", "urlHosts"):
        value = c.get(key)
        if value is None:
            continue
        if not isinstance(value, list) or any(not isinstance(e, str) for e in value):
            raise PolicyError(f'"{key}" must be an array of strings', path)

    for key in ("oneOf", "noneOf"):
        if c.get(key) is not None and not isinstance(c[key], list):
            raise PolicyError(f'"{key}" must be an array', path)

    for key in ("min", "max", "maxLength", "minLength"):
        if c.get(key) is not None and not _is_number(c[key]):
            raise PolicyError(f'"{key}" must be a number', path)

    optional = c.get("optional")
    if optional is not None and not isinstance(optional, bool):
        raise PolicyError('"optional" must be a boolean', path)

    type_ = c.get("type")
    if type_ is not None and type_ not in _ARG_TYPES:
        raise PolicyError(
            '"type" must be one of string, number, boolean, object, array', path
        )

    return ArgConstraint(**{key: c.get(key) for key in _CONSTRAINT_KEYS})


def cast_effect(value: Any) -> Effect:
    """Narrow an already-validated string to the Effect literal type."""
    assert value in EFFECTS
    return value  # type: ignore[return-value]


def _is_number(value: Any) -> bool:
    """JavaScript's ``typeof x === 'number'``: booleans are not numbers."""
    return isinstance(value, (int, float)) and not isinstance(value, bool)


def _is_integer(value: float) -> bool:
    return math.isfinite(value) and float(value).is_integer()


def _require_object(value: Any, path: str) -> dict[str, Any]:
    if not isinstance(value, dict):
        raise PolicyError("expected an object", path)
    return value


def _reject_unknown_keys(obj: dict[str, Any], allowed: tuple[str, ...], path: str) -> None:
    unknown = [key for key in obj if key not in allowed]
    if unknown:
        raise PolicyError(
            f"unknown field(s) {', '.join(js_json(k) for k in unknown)}; "
            f"allowed: {', '.join(allowed)}",
            path,
        )

File: src/policy/load.ts
/**
 * Policy parsing and validation.
 *
 * Validation is strict and rejects unknown keys. In a security policy a typo
 * is not a harmless no-op: `startWith` silently ignored would turn a scoped
 * filesystem rule into an unscoped one. Fail loudly at load time instead.
 */

import { readFileSync } from 'node:fs';
import { parse as parseYaml } from 'yaml';
import type { ArgConstraint, Policy, PolicyRule } from '../types.js';

export class PolicyError extends Error {
  constructor(message: string, readonly path: string) {
    super(`${path}: ${message}`);
    this.name = 'PolicyError';
  }
}

const EFFECTS = new Set(['allow', 'deny', 'ask']);
const POLICY_KEYS = new Set(['version', 'name', 'default', 'budget', 'plan', 'rules']);
const PLAN_KEYS = new Set(['purpose', 'approvedBy', 'warnAt']);
const RULE_KEYS = new Set(['id', 'description', 'tools', 'effect', 'when', 'limit']);
const BUDGET_KEYS = new Set(['calls', 'tokens', 'usd', 'bytes', 'seconds']);
const LIMIT_KEYS = new Set(['max', 'perSeconds']);
const CONSTRAINT_KEYS = new Set([
  'optional', 'oneOf', 'noneOf', 'matches', 'startsWith', 'excludes',
  'min', 'max', 'maxLength', 'minLength', 'urlHosts', 'type',
]);

/** Read and validate a policy from a `.yaml`, `.yml` or `.json` file. */
export function loadPolicyFile(file: string): Policy {
  let raw: string;
  try {
    raw = readFileSync(file, 'utf8');
  } catch (err) {
    throw new PolicyError(`cannot read policy file (${(err as Error).message})`, file);
  }
  return parsePolicy(raw, file);
}

/** Validate an already-read policy document. `source` only decorates errors. */
export function parsePolicy(raw: string, source = 'policy'): Policy {
  let doc: unknown;
  try {
    doc = parseYaml(raw);
  } catch (err) {
    throw new PolicyError(`invalid YAML/JSON (${(err as Error).message})`, source);
  }
  return validatePolicy(doc, source);
}

export function validatePolicy(doc: unknown, source = 'policy'): Policy {
  const root = requireObject(doc, source);
  rejectUnknownKeys(root, POLICY_KEYS, source);

  if (root.version !== '1') {
    throw new PolicyError(
      `unsupported policy version ${JSON.stringify(root.version)}; expected "1"`,
      source
    );
  }

  if (root.name !== undefined && typeof root.name !== 'string') {
    throw new PolicyError('"name" must be a string', source);
  }

  if (root.default !== undefined && !EFFECTS.has(root.default as string)) {
    throw new PolicyError('"default" must be one of allow, deny, ask', source);
  }

  const policy: Policy = {
    version: '1',
    rules: [],
    ...(root.name !== undefined ? { name: root.name as string } : {}),
    ...(root.default !== undefined ? { default: root.default as Policy['default'] } : {}),
  };

  if (root.budget !== undefined) {
    const budget = requireObject(root.budget, `${source}.budget`);
    rejectUnknownKeys(budget, BUDGET_KEYS, `${source}.budget`);
    for (const key of BUDGET_KEYS) {
      const value = budget[key];
      if (value === undefined) continue;
      if (typeof value !== 'number' || !Number.isFinite(value) || value <= 0) {
        throw new PolicyError(`"${key}" must be a positive number`, `${source}.budget`);
      }
    }
    policy.budget = budget as Policy['budget'];
  }

  if (root.plan !== undefined) {
    const plan = requireObject(root.plan, `${source}.plan`);
    rejectUnknownKeys(plan, PLAN_KEYS, `${source}.plan`);

    if (typeof plan.purpose !== 'string' || plan.purpose.trim().length === 0) {
      throw new PolicyError('"purpose" must be a non-empty string', `${source}.plan`);
    }
    if (plan.approvedBy !== undefined && typeof plan.approvedBy !== 'string') {
      throw new PolicyError('"approvedBy" must be a string', `${source}.plan`);
    }
    if (plan.warnAt !== undefined) {
      if (!Array.isArray(plan.warnAt) || plan.warnAt.length === 0) {
        throw new PolicyError('"warnAt" must be a non-empty array', `${source}.plan`);
      }
      for (const t of plan.warnAt) {
        if (typeof t !== 'number' || !(t > 0) || t > 1) {
          throw new PolicyError(
            '"warnAt" entries must be fractions greater than 0 and at most 1, e.g. 0.8',
            `${source}.plan`
          );
        }
      }
    }
    // A plan with nothing to measure against would report nothing, which reads
    // as "all clear" rather than "not configured". Fail loudly instead.
    if (root.budget === undefined) {
      throw new PolicyError(
        'a "plan" requires a "budget" to measure against; add budget limits or remove the plan',
        source
      );
    }

    policy.plan = {
      purpose: plan.purpose,
      ...(plan.approvedBy !== undefined ? { approvedBy: plan.approvedBy as string } : {}),
      ...(plan.warnAt !== undefined
        ? { warnAt: [...(plan.warnAt as number[])].sort((a, b) => a - b) }
        : {}),
    };
  }

  if (!Array.isArray(root.rules)) {
    throw new PolicyError('"rules" must be an array', source);
  }

  const seen = new Set<string>();
  policy.rules = root.rules.map((rule, i) => {
    const parsed = validateRule(rule, `${source}.rules[${i}]`);
    if (seen.has(parsed.id)) {
      throw new PolicyError(`duplicate rule id "${parsed.id}"`, `${source}.rules[${i}]`);
    }
    seen.add(parsed.id);
    return parsed;
  });

  return policy;
}

function validateRule(raw: unknown, path: string): PolicyRule {
  const rule = requireObject(raw, path);
  rejectUnknownKeys(rule, RULE_KEYS, path);

  if (typeof rule.id !== 'string' || rule.id.length === 0) {
    throw new PolicyError('"id" must be a non-empty string', path);
  }
  if (rule.description !== undefined && typeof rule.description !== 'string') {
    throw new PolicyError('"description" must be a string', path);
  }
  if (!Array.isArray(rule.tools) || rule.tools.length === 0) {
    throw new PolicyError('"tools" must be a non-empty array of glob patterns', path);
  }
  for (const tool of rule.tools) {
    if (typeof tool !== 'string' || tool.length === 0) {
      throw new PolicyError('every entry in "tools" must be a non-empty string', path);
    }
  }
  if (!EFFECTS.has(rule.effect as string)) {
    throw new PolicyError('"effect" must be one of allow, deny, ask', path);
  }

  const out: PolicyRule = {
    id: rule.id,
    tools: rule.tools as string[],
    effect: rule.effect as PolicyRule['effect'],
    ...(rule.description !== undefined ? { description: rule.description as string } : {}),
  };

  if (rule.when !== undefined) {
    const when = requireObject(rule.when, `${path}.when`);
    const constraints: Record<string, ArgConstraint> = {};
    for (const [argPath, constraint] of Object.entries(when)) {
      constraints[argPath] = validateConstraint(constraint, `${path}.when["${argPath}"]`);
    }
    out.when = constraints;
  }

  if (rule.limit !== undefined) {
    const limit = requireObject(rule.limit, `${path}.limit`);
    rejectUnknownKeys(limit, LIMIT_KEYS, `${path}.limit`);
    if (typeof limit.max !== 'number' || !Number.isInteger(limit.max) || limit.max < 0) {
      throw new PolicyError('"max" must be a non-negative integer', `${path}.limit`);
    }
    if (
      limit.perSeconds !== undefined &&
      (typeof limit.perSeconds !== 'number' || limit.perSeconds <= 0)
    ) {
      throw new PolicyError('"perSeconds" must be a positive number', `${path}.limit`);
    }
    out.limit = {
      max: limit.max,
      ...(limit.perSeconds !== undefined ? { perSeconds: limit.perSeconds as number } : {}),
    };
  }

  return out;
}

/**
 * Reject a regular expression whose shape can backtrack catastrophically.
 *
 * `matches` patterns are compiled once at load but RUN against argument values,
 * which are model output — the untrusted side. A pattern with a quantifier
 * applied to a group that already contains one, `^(([a-z]+)+@)+x$` and its
 * family, takes exponential time on a crafted 44-character value: over ten
 * seconds in both implementations, with the whole enforcement point frozen
 * because evaluation is synchronous.
 *
 * The check is STRUCTURAL rather than timed on purpose. Rejecting a pattern
 * because it ran slowly would make acceptance depend on the machine and the
 * regex engine, so the same policy could load in one implementation and fail in
 * the other — and "the same policy always reaches the same verdict" is the
 * property this library is for. A shape test gives both languages the same
 * answer everywhere.
 *
 * It is a heuristic, and the docs say so: it catches the classic nested-
 * quantifier family, not every pathological pattern. A linear-time engine is
 * the real fix and would cost this project its single-dependency property, so
 * it is a deliberate item on the roadmap rather than a silent gap.
 */
export function nestedQuantifier(pattern: string): string | null {
  /** Groups currently open, and whether a quantifier has been seen inside each. */
  const stack: { quantified: boolean }[] = [];
  let inClass = false;

  for (let i = 0; i < pattern.length; i++) {
    const ch = pattern[i]!;

    if (ch === '\\') {
      i++; // An escaped character is a literal, never syntax.
      continue;
    }
    if (inClass) {
      if (ch === ']') inClass = false;
      continue;
    }
    if (ch === '[') {
      inClass = true;
      continue;
    }
    if (ch === '(') {
      stack.push({ quantified: false });
      continue;
    }
    if (ch === ')') {
      const closed = stack.pop();
      if (!closed) continue; // Unbalanced; the RegExp constructor will complain.

      // A quantifier directly after this group, when the group itself contained
      // one, is the catastrophic shape.
      const next = quantifierAt(pattern, i + 1);
      if (next !== null) {
        if (closed.quantified) {
          return `nested quantifier: a "${next}" applied to a group that already contains a quantifier`;
        }
        // The group is quantified, so the group it sits inside now counts as
        // containing a quantifier too.
        if (stack.length > 0) stack[stack.length - 1]!.quantified = true;
      }
      continue;
    }
    if (isQuantifier(pattern, i) && stack.length > 0) {
      stack[stack.length - 1]!.quantified = true;
    }
  }

  return null;
}

/** The quantifier token starting at `index`, or null if there is not one. */
function quantifierAt(pattern: string, index: number): string | null {
  const ch = pattern[index];
  if (ch === '*' || ch === '+') return ch;
  if (ch === '{') {
    const close = pattern.indexOf('}', index);
    if (close === -1) return null;
    const body = pattern.slice(index + 1, close);
    // Only an open-ended or plural repetition can blow up; {0,1} cannot.
    if (!/^\d*,?\d*$/.test(body)) return null;
    const [min, max] = body.split(',');
    const upper = max === undefined ? Number(min) : max === '' ? Infinity : Number(max);
    return upper > 1 ? `{${body}}` : null;
  }
  return null;
}

function isQuantifier(pattern: string, index: number): boolean {
  return quantifierAt(pattern, index) !== null;
}

function validateConstraint(raw: unknown, path: string): ArgConstraint {
  const c = requireObject(raw, path);
  rejectUnknownKeys(c, CONSTRAINT_KEYS, path);

  if (c.matches !== undefined) {
    if (typeof c.matches !== 'string') throw new PolicyError('"matches" must be a string', path);
    try {
      new RegExp(c.matches);
    } catch (err) {
      throw new PolicyError(`"matches" is not a valid regexp (${(err as Error).message})`, path);
    }
    const unsafe = nestedQuantifier(c.matches);
    if (unsafe !== null) {
      throw new PolicyError(
        `"matches" has a shape that can backtrack catastrophically — ${unsafe}. ` +
          `Argument values come from the model, so a crafted value would freeze ` +
          `enforcement. Prefer "startsWith", "oneOf" or "excludes", or rewrite the ` +
          `pattern without the nesting`,
        path
      );
    }
  }

  for (const key of ['startsWith', 'excludes', 'urlHosts'] as const) {
    const value = c[key];
    if (value === undefined) continue;
    if (!Array.isArray(value) || value.some((entry) => typeof entry !== 'string')) {
      throw new PolicyError(`"${key}" must be an array of strings`, path);
    }
  }

  for (const key of ['oneOf', 'noneOf'] as const) {
    if (c[key] !== undefined && !Array.isArray(c[key])) {
      throw new PolicyError(`"${key}" must be an array`, path);
    }
  }

  for (const key of ['min', 'max', 'maxLength', 'minLength'] as const) {
    if (c[key] !== undefined && typeof c[key] !== 'number') {
      throw new PolicyError(`"${key}" must be a number`, path);
    }
  }

  if (c.optional !== undefined && typeof c.optional !== 'boolean') {
    throw new PolicyError('"optional" must be a boolean', path);
  }

  if (
    c.type !== undefined &&
    !['string', 'number', 'boolean', 'object', 'array'].includes(c.type as string)
  ) {
    throw new PolicyError(
      '"type" must be one of string, number, boolean, object, array',
      path
    );
  }

  return c as ArgConstraint;
}

function requireObject(value: unknown, path: string): Record<string, unknown> {
  if (typeof value !== 'object' || value === null || Array.isArray(value)) {
    throw new PolicyError('expected an object', path);
  }
  return value as Record<string, unknown>;
}

function rejectUnknownKeys(obj: Record<string, unknown>, allowed: Set<string>, path: string): void {
  const unknown = Object.keys(obj).filter((key) => !allowed.has(key));
  if (unknown.length > 0) {
    throw new PolicyError(
      `unknown field(s) ${unknown.map((k) => `"${k}"`).join(', ')}; allowed: ${[...allowed].join(', ')}`,
      path
    );
  }
}
